Add separate scale and offset for current temperature for modbus climates

illia-piskurov opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The documentation previously stated that scale and offset only applied to the target temperature.
This was incorrect: in reality they apply to both target and current temperature.

This PR adds the ability to configure separate scale and offset for current and target temperatures.

There are devices that require different scaling for target and current temperatures (for example, Cooper air conditioners). People have also reported this limitation on the forum:

https://community.home-assistant.io/t/wth-why-do-modbus-climate-entities-only-have-one-scale-setting-for-temperature-registers/803414

https://community.home-assistant.io/t/modbus-plataform-climate/323571

This was also discussed in PR: home-assistant/core#135848.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://community.home-assistant.io/t/wth-why-do-modbus-climate-entities-only-have-one-scale-setting-for-temperature-registers/803414
- This PR is related to issue: https://community.home-assistant.io/t/modbus-plataform-climate/323571
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40770
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

illia-piskurov committed:
Support of current temp scale and offset

janiversen reviewed (changes requested):
You are adding 2 new config parameters, that seems to compete with the existing scale/offset, and I cannot find the relationship ???

You remove _attr_current_temperature, so I wonder how the UI knows about the current temperature.

janiversen committed:
Merge branch 'dev' into curr_temp_scale_and_offset

illia-piskurov commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -212,6 +214,8 @@
         vol.Optional(CONF_STRUCTURE): cv.string,
         vol.Optional(CONF_SCALE, default=1): vol.Coerce(float),
         vol.Optional(CONF_OFFSET, default=0): vol.Coerce(float),
+        vol.Inclusive(CONF_CURRENT_TEMP_SCALE, "current_temp"): vol.Coerce(float),
```
If both are configured, CONF_SCALE/CONF_OFFSET are still applied to the target temperature, while CONF_CURRENT_TEMP_SCALE/CONF_CURRENT_TEMP_OFFSET are only applied to the current temperature.

This separation was introduced because some devices expose current and target temperature registers with different scaling/offset requirements.

Using the dedicated options ensures flexibility, while keeping the old behavior intact: if CONF_CURRENT_TEMP_SCALE/OFFSET are not set, the integration continues to use CONF_SCALE/OFFSET for both values, so existing configurations are not broken.

illia-piskurov replied:
First I’ll fix the actual bug with scale/offset being applied to both target and current temperature – that should be sorted out first.

About the config: I see how it’s confusing to have CONF_CURRENT_* available everywhere. I’ll limit these options to climate only, since that’s where the use case really is (current vs. target temperature registers). Sensors like “water volume” won’t be affected.

The idea is:

By default, CONF_SCALE/CONF_OFFSET still apply to both target and current temperature, so nothing changes for existing configs.

If you explicitly set CONF_CURRENT_TEMP_SCALE/OFFSET, those override the scaling for the current temperature only. Target temperature continues to use the normal scale/offset.

The main goal here is not to break old configurations, so if you don’t touch the new options, behavior stays exactly the same.

As for adding also CONF_TARGET_TEMP_*: sure, that would be cleaner symmetry, but implementing it properly would mean a bigger refactor, since scaling is currently handled centrally in __process_raw_value in entity.py. With this change I’m keeping it simple: just undo the global scale/offset for current temperature and apply the custom pair. Later we could discuss a proper per-register pipeline that would allow target-specific options too.

crug80 replied:
By adding these new parameters at the base_struct it means that also a sensor (not only a climate) could have configured those parameters. 
Why can't you add them to the CLIMATE_SCHEMA, instead?

illia-piskurov commented on homeassistant/components/modbus/climate.py:
```diff
@@ -474,9 +483,17 @@ async def _async_update(self) -> None:
         )
 
-        self._attr_current_temperature = await self._async_read_register(
+        current_temp_register_value = await self._async_read_register(
```
in 490 or 495 line

illia-piskurov commented on homeassistant/components/modbus/climate.py:
```diff
@@ -301,6 +303,13 @@ def __init__(
         else:
             self._hvac_onoff_coil = None
 
+        if CONF_CURRENT_TEMP_SCALE in config:
```
I followed the same style used in the surrounding code for consistency. If you prefer, I can change it to a one-liner with .get, but I kept it aligned with the existing pattern.

illia-piskurov replied:
line 172, 203, 231, 267, 289, 299

illia-piskurov commented on homeassistant/components/modbus/climate.py:
```diff
@@ -474,9 +483,17 @@ async def _async_update(self) -> None:
+            # Undo global scale/offset
+            raw_value = (current_temp_register_value - self._offset) / self._scale
+            new_value = raw_value * self._current_temp_scale + self._current_temp_offset
+            self._attr_current_temperature = new_value
```
I don’t understand how to bypass unpack_structure_result, which calls __process_raw_value that always applies scale and offset.

illia-piskurov committed:
Try another way

janiversen commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -273,6 +277,10 @@
             vol.Optional(CONF_TEMPERATURE_UNIT, default=DEFAULT_TEMP_UNIT): cv.string,
             vol.Exclusive(CONF_HVAC_ONOFF_COIL, "hvac_onoff_type"): cv.positive_int,
             vol.Exclusive(CONF_HVAC_ONOFF_REGISTER, "hvac_onoff_type"): cv.positive_int,
+            vol.Inclusive(CONF_CURRENT_TEMP_SCALE, "current_temp"): vol.Coerce(float),
```
Why is it not allowed just to add scale or offset, why does it need to be both.

If I remember correctly, you only needed scale but someone else asked for offset, so it is clear that some devices do not need both scale and offset.

illia-piskurov replied:
My position is that if someone tells me how to do it correctly, I will do it faster than sending 30 variants that will be rejected. And you will save time by not having to check incorrect solutions.

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -301,6 +305,20 @@ def __init__(
+            self._current_temp_scale = None
+            self._current_temp_offset = None
+
+        if CONF_TARGET_TEMP_SCALE in config and CONF_TARGET_TEMP_OFFSET in config:
```
This looks wrong, you should be able to use scale without offset.

And please use .get as I remarked earlier.

illia-piskurov replied:
I don't understand. In lines 174, 205, 233, 269, etc., .get is not used, but instead it is written in the style of “in config”. Why should I have it in a different format?

janiversen replied:
I get the distinct feeling you do not want my review comments ! Feel free to ask for a review by someone else.

I am getting into a "why" discussion, that is NOT the purpose of a review.

illia-piskurov replied:
I just thought everything should be in the same style. As usual in “style guides.”

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -412,9 +430,14 @@ async def async_set_swing_mode(self, swing_mode: str) -> None:
-        target_temperature = (
-            float(kwargs[ATTR_TEMPERATURE]) - self._offset
-        ) / self._scale
+        if self._target_temp_scale is not None and self._target_temp_offset is not None:
```
this should be "or" not and....


but with this PR we are now trailing 3 times scale/offset through the component, that does not seem right, it would be a lot easier to eliminate the old scale/offset, by assigning those to target_ and current_ where you validate the parameters.

janiversen replied:
Exactly as I wrote !! Do you seriously want me to write the code ?

janiversen replied:
That would be the same as writing the code, it is your PR so it is your decision, there are no rules so you might choose another way that I would choose, which is perfectly OK.

I review what you suggest, and I have already told you what I look out for.

janiversen reviewed (changes requested):
I still miss a validation of scale/offset vs current_, it seems I can use both without knowing how it is solved.

I still miss the correction of the bug you mention in the text !

Please make sure you solve all comments before requesting a review, not doing so is wasting my time.

janiversen commented:
The PR text goes out to all users, so please be specific what is not working in the code ? (if the documentation is incorrect you should open a PR in the documentation repo).

When there is a bug in the code, it is important to inform users the specifics, so the6 can determine how it might have affected them.

illia-piskurov committed:
Separated offset and scale

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -412,9 +420,10 @@ async def async_set_swing_mode(self, swing_mode: str) -> None:
-            float(kwargs[ATTR_TEMPERATURE]) - self._offset
-        ) / self._scale
+        scale = self._target_temp_scale or self._scale
+        offset = self._target_temp_offset or self._offset
```
same here.

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -472,11 +481,17 @@ async def _async_update(self) -> None:
             self._target_temperature_register[
                 HVACMODE_TO_TARG_TEMP_REG_INDEX_ARRAY[self._attr_hvac_mode]
             ],
+            scale=self._target_temp_scale or self._scale,
```
same here.

janiversen commented on homeassistant/components/modbus/entity.py:
```diff
@@ -256,16 +256,29 @@ def _swap_registers(self, registers: list[int], slave_count: int) -> list[int]:
+        entry: float | str | bytes,
+        scale: float | None = None,
+        offset: float | None = None,
+    ) -> str | None:
```
this is an internal method, so why allow None ??? 

You already substitute None in the calling method.

janiversen reviewed (changes requested):
Please solve the comments as well as the mypy problem.

I could not find the validation, that temp_offset is not used simultaneously with offset ?

I also cannot find the code you made to justify the comment in the PR text:

""Now scale and offset is applied to both target and current temperature at the same time. This is an error in the documentation:
https://www.home-assistant.io/integrations/modbus/#scale""

Please give me a pointer to where you solved this, without using the new parameters, if it is a bug is should be solved !

illia-piskurov committed:
Removed mistakes

illia-piskurov committed:
Change in sensor.py

illia-piskurov commented:
> Please solve the comments as well as the mypy problem.
> 
> I could not find the validation, that temp_offset is not used simultaneously with offset ?
> 
> I also cannot find the code you made to justify the comment in the PR text:
> 
> ""Now scale and offset is applied to both target and current temperature at the same time. This is an error in the documentation: https://www.home-assistant.io/integrations/modbus/#scale""
> 
> Please give me a pointer to where you solved this, without using the new parameters, if it is a bug is should be solved !

In the documentation it is stated that scale and offset are explicitly applied only to the target temperature. But in reality, in the current code they are also applied to the current temperature.

The idea of my PR is to make this behavior configurable: to allow separate scale and offset settings for both target and current temperature. At the same time, the old behavior is preserved for backward compatibility.

Once this PR is accepted, I will also update the documentation to reflect this new information.

janiversen commented:
> > ""Now scale and offset is applied to both target and current temperature at the same time. This is an error in the documentation: https://www.home-assistant.io/integrations/modbus/#scale""
> > Please give me a pointer to where you solved this, without using the new parameters, if it is a bug is should be solved !
> 
> In the documentation it is stated that scale and offset are explicitly applied only to the target temperature. But in reality, in the current code they are also applied to the current temperature.

Since you mentions it here, any user reading it would expect the code to be corrected, which you have not done.

Alternative is of course to make a documentation pull request with a correction, and reference that in this PR...that way a user would not expect the code to be changed.

> The idea of my PR is to make this behavior configurable: to allow separate scale and offset settings for both target and current temperature. At the same time, the old behavior is preserved for backward compatibility.

Not only !!! if that was the sole idea, you would not have written about the bug.

> Once this PR is accepted, I will also update the documentation to reflect this new information.

Actually that need to happen before, we never accept a code PR with parameter changes without a corresponding Documentation PR....however that can be done as the last step.

janiversen reviewed (changes requested):
Please excuse me for asking, but in the view of my last review, did you take the time to actually look at code, or did you just program until it worked ?

All the review comments this time, is somehow I have written earlier, so I expected you to be observant of that.

The type mess surprises me a lot, that is so obvious wrong, when looking at the code.

janiversen commented:
Did you read my comment after my last review:

"Please do not ask me to do more reviews, unless you are absolutely sure everything is solved and ok !"

I think it is better if you do not re-request a review from me, since you obviously do not want to listen to my review comments.

@crug80 you might want to chime in here, I have spent far too much time trying to help....and we still at obvious things like consistent typing of scale/offset.

illia-piskurov commented:
> since you obviously do not want to listen to my review comments.

I'm doing everything you asked me to fix (like with .get, ditching the scale and offset cancellation, and making the changes below). And for typing, I went with what mypy and ruff suggest in the pre-commit hook.

illia-piskurov commented:
> And not to forget, you have consistently asked for review, just to get an opinion or even though mypy was failing.

Because I am not a senior Python developer, I simply need this feature, just like other users. I am doing what I can. I need help.

janiversen commented:
@crug80 can you help the man ? I do not have time to tutor basic python.

crug80 commented:
@illia-piskurov, please check discord, so we can talk about this without wasting the PR.

crug80 commented:
Meanwhile I was waiting for a contact on discord, I wrote a[ ready to use PR](https://github.com/crug80/core/tree/scalesandoffsets), aligned with the current dev branch.
Let me know

crug80 committed:
Merge branch 'home-assistant:dev' into curr_temp_scale_and_offset

crug80 committed:
rev2-crug

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

illia-piskurov commented:
Added docs: https://github.com/home-assistant/home-assistant.io/pull/40768

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

janiversen commented:
The documentation change is ok, but the PR test seems wrong:

"Now scale and offset is applied to both target and current temperature at the same time"

as far as I can see, scale and offset are applied to target and current temperature on dev, so what have changed in code "now" signals a change, but if the code have not changed then please reword it accordingly.

It is correct that the documentation was incorrect, and it is now corrected.

Once the PR text is updated, ping me as reviewer and I will approve the PR (remembering the code PR is still open).

janiversen commented:
Before I start reviewing, I read the PR text:

"Now scale and offset is applied to both target and current temperature at the same time"

I think that is incorrect, looking at dev, it is already applied to both ! It is correct that the documentation is wrong. The "now" implies a change in this PR which I expect is not the case.

janiversen reviewed (changes requested):
looks better, but I do see problems due to the use of legal default values.

illia-piskurov commented:
> You added:
> 
> "This fix corrects the documentation."
> 
> How can a code PR fix the documentation ?

Fixed, updated the sentence to reflect the code changes.

janiversen committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,60 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+        )
+
+    if config[CONF_SCALE] != DEFAULT_SCALE:
+        config[CONF_TARGET_TEMP_SCALE] = config[CONF_SCALE]
```
It is not problem, because 1 is equal to 1.0. Anyway, I'll add a test where some of the parameters have the float notation 1.0 and some other not.

janiversen commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,60 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+def ensure_and_check_conflicting_scales_and_offsets(config: dict[str, Any]) -> dict:
+    """Check for conflicts in scale/offset and ensure target/current temp scale/offset is set."""
+
+    if (
```
It is not a technical problem, but it is still confusing for the user (and quite easy to avoid).

janiversen commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,60 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+        config[CONF_CURRENT_TEMP_OFFSET] = config[CONF_OFFSET]
+
+    return config
+
```
As I wrote to ensure that no part of the code continue to depend on them.....it is called robust programming.

janiversen commented on tests/components/modbus/test_init.py:
```diff
@@ -1423,3 +1431,75 @@ async def test_pb_service_write_no_slave(
+                CONF_TARGET_TEMP: 120,
+                CONF_ADDRESS: 117,
+                CONF_SLAVE: 10,
+                CONF_SCAN_INTERVAL: 0,
```
And I happen to disagree...if you do not delete them....then at least all places in the code (apart from schema and validator) should not reference CONF_SCALE !

Leaving references to CONF_SCALE in the code, is a bug waiting to happen, and the best way to ensure that is to remove CONF_SCALE from the config.

crug80 committed:
floats

janiversen commented:
Please stop requesting answers for every change you make. Solve all the review comments and mark it as ready for review.

If I have to respond to every step, I will not be able to review it.

crug80 commented on homeassistant/components/modbus/climate.py:
```diff
@@ -562,7 +576,12 @@ async def _async_update(self) -> None:
+        self,
+        register_type: str,
+        register: int,
+        scale: float = 1,
```
I don't like it, but done.

crug80 committed:
fix requests

crug80 commented:
> Please stop requesting answers for every change you make. Solve all the review comments and mark it as ready for review.
> 
> If I have to respond to every step, I will not be able to review it.

The answers helped me to understand the nature of your requests, hopefully.

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

janiversen commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,51 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+
+    if CONF_TARGET_TEMP_SCALE not in config:
+        config[CONF_TARGET_TEMP_SCALE] = config.get(CONF_SCALE, DEFAULT_SCALE)
+    elif CONF_SCALE in config and config[CONF_TARGET_TEMP_SCALE] != config[CONF_SCALE]:
```
No !!!! scale and target/current are not to be used simultaneously.

janiversen replied:
This goes right down all the checks.

This method could be simplified quite a lot.

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -413,8 +421,8 @@ async def async_set_swing_mode(self, swing_mode: str) -> None:
-            float(kwargs[ATTR_TEMPERATURE]) - self._offset
-        ) / self._scale
+            float(kwargs[ATTR_TEMPERATURE]) - self._target_temp_offset
+        ) / self._target_temp_scale
```
Do you validate for target/current scale being configured to 0.

It does not make sense, but it is possible.

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -472,15 +480,21 @@ async def _async_update(self) -> None:
-            self._input_type, self._address
+            self._input_type,
+            self._address,
+            scale=self._current_temp_scale,
```
This method only have positional parameters !

janiversen commented on homeassistant/components/modbus/entity.py:
```diff
@@ -181,8 +177,6 @@ def __init__(self, hass: HomeAssistant, hub: ModbusHub, config: dict) -> None:
         else:
             self._precision = config.get(CONF_PRECISION, 0)
-            if self._precision > 0 or self._scale != int(self._scale):
-                self._value_is_int = False
```
Removing precision here, might be correct, but it seems it is not added to climate ??

janiversen commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -118,7 +118,9 @@ async def _async_update(self) -> None:
         self._attr_available = True
-        result = self.unpack_structure_result(raw_result.registers)
+        result = self.unpack_structure_result(
+            raw_result.registers, self._scale, self._offset
```
This is still a problem.

janiversen replied:
??

janiversen commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,51 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+            f"Invalid offsets: {CONF_OFFSET} and {CONF_CURRENT_TEMP_OFFSET} cannot be used together, please use only one of them."
+        )
+
+    if CONF_OFFSET in config:
```
And what happens if a sensor uses offset ?

That was not what I wrote. Do NOT have both scale and target/current, that can lead to problems.

crug80 replied:
With regards to offset, the validator works on climate only, so no problems for a sensor. 
With regards to the _offset and _scale, they belong to the sensor only. 
I moved them out from BaseStructPlatform to ModbusSensor, so they cannot be called or used inside the climate.

janiversen reviewed (changes requested):
Please do not request a review, when some of the review comments have not been attended to.

crug80 committed:
fix requests 2

crug80 committed:
Merge branch 'home-assistant:dev' into curr_temp_scale_and_offset

janiversen committed:
Merge branch 'dev' into curr_temp_scale_and_offset

janiversen commented on tests/components/modbus/test_climate.py:
```diff
@@ -1699,3 +1705,147 @@ async def test_no_discovery_info_climate(
+    await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_ID)
+    assert state.attributes.get("temperature") == result
```
Seems there are no x_scale: 0 tests ?

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -499,7 +513,7 @@ async def _async_update(self) -> None:
         if self._hvac_action_register is not None:
             hvac_action = await self._async_read_register(
-                self._hvac_action_type, self._hvac_action_register, raw=True
+                self._hvac_action_type, self._hvac_action_register, 1.0, 0, raw=True
```
Whe define constants, when not using them ?

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -517,6 +531,8 @@ async def _async_update(self) -> None:
                 if isinstance(self._fan_mode_register, int)
                 else self._fan_mode_register[0],
+                1.0,
+                0,
```
same

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -533,6 +549,8 @@ async def _async_update(self) -> None:
                 self._swing_mode_register
                 if isinstance(self._swing_mode_register, int)
                 else self._swing_mode_register[0],
+                1.0,
```
same

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -551,7 +569,7 @@ async def _async_update(self) -> None:
         if self._hvac_onoff_register is not None:
             onoff = await self._async_read_register(
-                CALL_TYPE_REGISTER_HOLDING, self._hvac_onoff_register, raw=True
+                CALL_TYPE_REGISTER_HOLDING, self._hvac_onoff_register, 1.0, 0, raw=True
```
same

janiversen commented on homeassistant/components/modbus/entity.py:
```diff
@@ -206,7 +200,12 @@ def _swap_registers(self, registers: list[int], slave_count: int) -> list[int]:
+    def __process_raw_value(
+        self,
+        entry: float | bytes,
+        scale: float = 1.0,
```
same here.

janiversen commented on homeassistant/components/modbus/entity.py:
```diff
@@ -226,7 +225,9 @@ def __process_raw_value(self, entry: float | str | bytes) -> str | None:
 
-    def unpack_structure_result(self, registers: list[int]) -> str | None:
+    def unpack_structure_result(
+        self, registers: list[int], scale: float = 1.0, offset: float = 0
```
and here.

janiversen reviewed (dismissed):
Seems a lot of the comments are about standard python, like do not use numeric constants in the code.

janiversen commented:
It means, the review comment was left open, not a single comment, and no obvious action.

crug80 commented:
> It means, the review comment was left open, not a single comment, and no obvious action.

Uhm.. I have already answered about that: https://github.com/home-assistant/core/pull/150985#discussion_r2333819326

crug80 committed:
fix requests 3

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 committed:
fix ruff

illia-piskurov committed:
Merge branch 'dev' into curr_temp_scale_and_offset

illia-piskurov committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

richardstone commented:
Hi @crug80 , @janiversen ,
I'm faced with the very same issue @dusan-ivanco described above. I don't really get it why some manufacturers do offset on one temp, and no offset on the other, but it's very annoying:
- I'm seeing temp of `89 C` on the measured but the set_temp is correct. 
- Or I set the offset, then the measured temp is correct while set_temp is `-9 C`

Any chance this will get merged soon?
Thanks a lot!

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

frenck committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 committed:
Update homeassistant/components/modbus/validators.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

illia-piskurov committed:
Merge branch 'dev' into curr_temp_scale_and_offset

emontnemery commented on homeassistant/components/modbus/validators.py:
```diff
@@ -243,6 +251,37 @@ def duplicate_fan_mode_validator(config: dict[str, Any]) -> dict:
+        if target_key not in config or config[target_key] == 0:
+            config[target_key] = default_value
+        if current_key not in config or config[current_key] == 0:
+            config[current_key] = default_value
```
OK. I'd suggest to remove the `or config[current_key] == 0` (as Frenck suggested) because it's confusing when reading the code and prone to break (for example if the default changes in the future), and instead add an additional validator to the scales:
```py
def not_zero(val: float) -> float:
    if val == 0:
        raise vol.Invalid("blablabla")
```

```diff
-        vol.Optional(CONF_SCALE): vol.Coerce(float),
+        vol.Optional(CONF_SCALE): vol.All(vol.Coerce(float), not_zero))
```

Or is there some reason to allow the user to set the scale to 0?

crug80 replied:
There is no reason to have 0 as scale, but instead you'll have an error in case you set due to division by 0. I'll change the code as per your suggestion. Thanks Erik

crug80 committed:
Merge branch 'home-assistant:dev' into curr_temp_scale_and_offset

crug80 committed:
fix  emontnemery request

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

crug80 committed:
Merge branch 'dev' into curr_temp_scale_and_offset

emontnemery committed:
Merge branch 'dev' into curr_temp_scale_and_offset

emontnemery reviewed (approved):
LGTM, thanks @illia-piskurov and @crug80 👍
